src/raftstore/store/fsm/store.rs

```diff
@@ -740,370 +1331,257 @@ impl<T: Transport, C: PdClient> Store<T, C> {
-                                key.region_id, key
-                            );
-                            self.snap_mgr.delete_snapshot(&key, s.as_ref(), false);
-                        }
+                    for (key, is_sending) in snaps {
+                        let snap = if is_sending {
+                            self.ctx.snap_mgr.get_snapshot_for_sending(&key)?
+                        } else {
+                            self.ctx.snap_mgr.get_snapshot_for_applying(&key)?
+                        };
+                        self.ctx.snap_mgr.delete_snapshot(&key, snap, false);
                     }
                 }
-            } else if key.term <= compacted_term
-                && (key.idx < compacted_idx || key.idx == compacted_idx && !is_applying_snap)
-            {
-                info!(
-                    "[region {}] snap file {} has been applied, delete.",
-                    key.region_id, key
-                );
-                let a = self.snap_mgr.get_snapshot_for_applying(&key)?;
-                self.snap_mgr.delete_snapshot(&key, a.as_ref(), false);
+                Err(TrySendError::Full(_)) => {}
+                Err(TrySendError::Disconnected(_)) => unreachable!(),
+            }
+        };
+        for (key, is_sending) in snap_keys {
+            if last_region_id == key.region_id {
+                keys.push((key, is_sending));
+                continue;
+            }
+
+            if !keys.is_empty() {
+                schedule_gc_snap(last_region_id, keys)?;
+                keys = vec![];
             }
+
+            last_region_id = key.region_id;
+            keys.push((key, is_sending));
+        }
+        if !keys.is_empty() {
+            schedule_gc_snap(last_region_id, keys)?;
         }
         Ok(())
     }
 
-    fn on_snap_mgr_gc(&mut self, event_loop: &mut EventLoop<Self>) {
+    fn on_snap_mgr_gc(&mut self) {
         if let Err(e) = self.handle_snap_mgr_gc() {
             error!("{} failed to gc snap manager: {:?}", self.tag, e);
         }
-        self.register_snap_mgr_gc_tick(event_loop);
+        self.register_snap_mgr_gc_tick();
     }
 
-    fn on_compact_lock_cf(&mut self, event_loop: &mut EventLoop<Self>) {
+    fn on_compact_lock_cf(&mut self) {
         // Create a compact lock cf task(compact whole range) and schedule directly.
-        if self.store_stat.lock_cf_bytes_written > self.cfg.lock_cf_compact_bytes_threshold.0 {
-            self.store_stat.lock_cf_bytes_written = 0;
+        let lock_cf_bytes_written = self
+            .ctx
+            .global_stat
+            .stat
+            .lock_cf_bytes_written
+            .load(Ordering::Relaxed);
+        if lock_cf_bytes_written > self.ctx.cfg.lock_cf_compact_bytes_threshold.0 {
+            self.ctx
+                .global_stat
+                .stat
+                .lock_cf_bytes_written
+                .fetch_sub(lock_cf_bytes_written, Ordering::Relaxed);
```

Comment:
How about just set to 0? Because the compact will be triggered after setting to 0.

Reply:
I prefer to keep the same old way. In the past, when writing lock cf twice of threshold, it may causes two times of compaction.